File: windows/MoDi.App.Contracts/Connectivity/SessionRegistry.cs
namespace MoDi.App.Contracts.Connectivity;

public sealed record SessionSnapshot(
    SessionId Id,
    TransportKind Transport,
    SessionState State,
    long Revision);

public readonly record struct SessionObservationResult(bool Applied, string? Reason = null);

/// <summary>
/// Observation-only registry used to compare the target session model with legacy runtime facts.
/// It deliberately exposes no connection or data-plane controls.
/// </summary>
public sealed class SessionRegistry
{
    private readonly object _gate = new();
    private readonly Dictionary<SessionId, SessionSnapshot> _sessions = [];
    private long _revision;

    public IReadOnlyList<SessionSnapshot> Snapshots
    {
        get
        {
            lock (_gate)
                return _sessions.Values.OrderBy(snapshot => snapshot.Id.Value, StringComparer.Ordinal).ToArray();
        }
    }

    public SessionObservationResult ObserveStarted(SessionId id, TransportKind transport, SessionState state)
    {
        lock (_gate)
        {
            if (_sessions.TryGetValue(id, out var existing))
            {
                if (existing.Transport != transport)
                    return new(false, "Session transport changed.");
                return ObserveStateLocked(existing, state);
            }

            _sessions[id] = new(id, transport, state, ++_revision);
            return new(true);
        }
    }

    public SessionObservationResult ObserveState(SessionId id, SessionState state)
    {
        lock (_gate)
        {
            if (!_sessions.TryGetValue(id, out var existing))
                return new(false, "Unknown session.");
            return ObserveStateLocked(existing, state);
        }
    }

    public SessionObservationResult ObserveEnded(SessionId id)
    {
        lock (_gate)
        {
            if (!_sessions.TryGetValue(id, out var existing))
                return new(false, "Unknown session.");
            if (existing.State == SessionState.Closed)
                return new(true);

            _sessions[id] = existing with { State = SessionState.Closed, Revision = ++_revision };
            return new(true);
        }
    }

    public void ObserveClosedAll()
    {
        lock (_gate)
        {
            foreach (var pair in _sessions.ToArray())
            {
                if (pair.Value.State != SessionState.Closed)
                    _sessions[pair.Key] = pair.Value with { State = SessionState.Closed, Revision = ++_revision };
            }
        }
    }

    private SessionObservationResult ObserveStateLocked(SessionSnapshot existing, SessionState state)
    {
        if (existing.State == state)
            return new(true);
        if (!SessionStateMachine.CanTransition(existing.State, state))
            return new(false, $"Illegal transition {existing.State} -> {state}.");

        _sessions[existing.Id] = existing with { State = state, Revision = ++_revision };
        return new(true);
    }
}

File: android/app/src/main/java/com/modi/connect/core/connectivity/SessionRegistry.kt
package com.modi.connect.core.connectivity

data class SessionSnapshot(
    val id: SessionId,
    val transport: TransportKind,
    val state: SessionState,
    val revision: Long,
)

data class SessionObservationResult(val applied: Boolean, val reason: String? = null)

/** Observation-only registry. It intentionally owns no connection or data-plane controls. */
class SessionRegistry {
    private val lock = Any()
    private val sessions = linkedMapOf<SessionId, SessionSnapshot>()
    private var revision = 0L

    val snapshots: List<SessionSnapshot>
        get() = synchronized(lock) { sessions.values.sortedBy { it.id.value }.toList() }

    fun observeStarted(id: SessionId, transport: TransportKind, state: SessionState): SessionObservationResult =
        synchronized(lock) {
            val existing = sessions[id]
            if (existing == null) {
                sessions[id] = SessionSnapshot(id, transport, state, ++revision)
                SessionObservationResult(true)
            } else if (existing.transport != transport) {
                SessionObservationResult(false, "Session transport changed.")
            } else {
                observeStateLocked(existing, state)
            }
        }

    fun observeState(id: SessionId, state: SessionState): SessionObservationResult = synchronized(lock) {
        val existing = sessions[id] ?: return@synchronized SessionObservationResult(false, "Unknown session.")
        observeStateLocked(existing, state)
    }

    fun observeEnded(id: SessionId): SessionObservationResult = synchronized(lock) {
        val existing = sessions[id] ?: return@synchronized SessionObservationResult(false, "Unknown session.")
        if (existing.state != SessionState.Closed) {
            sessions[id] = existing.copy(state = SessionState.Closed, revision = ++revision)
        }
        SessionObservationResult(true)
    }

    fun observeClosedAll() = synchronized(lock) {
        sessions.toMap().forEach { (id, snapshot) ->
            if (snapshot.state != SessionState.Closed) {
                sessions[id] = snapshot.copy(state = SessionState.Closed, revision = ++revision)
            }
        }
    }

    private fun observeStateLocked(existing: SessionSnapshot, state: SessionState): SessionObservationResult {
        if (existing.state == state) return SessionObservationResult(true)
        if (!SessionStateMachine.canTransition(existing.state, state)) {
            return SessionObservationResult(false, "Illegal transition ${existing.state} -> $state.")
        }
        sessions[existing.id] = existing.copy(state = state, revision = ++revision)
        return SessionObservationResult(true)
    }
}
